### PR TITLE
[TIR] cast disparate floating point types for binary ops

### PR DESCRIPTION
**Pull request by AndrewZhaoLuo**
Right now if we in TIR add a floating point type with an integer type, the integer type will be implicitly cast to the floating point type.

E.g. `a: float32 + b: int32 ---> a: float32 + cast(float32, b: int32)`

This change does the same thing for floating point types. If we have two different floating point types e.g. fp16 and fp32, then when operating, the lower bit floating point type gets cast into the higher bit floating point type.

E.g. `a: float32 + b: float16 --> a: float32 + cast(float32, b: float16)`

This is of use since https://github.com/apache/tvm/issues/8340 has an issue where some schedules which should support mixed precision types do not. This is due to binary ops like addition and multiplication not supporting mixing fp32 and fp16. Most of these errors can be fixed by inserting a cast into the schedule. 

Rather than manually audit every schedule which might have this, this might be a preferable and reasonable solution.

### CODE REVIEW

File: src/tir/op/op.cc
Comment (comaniac) [inline src/tir/op/op.cc:123]: Better to rewire an entire comment here. Here is my thought and feel free to use your own words:
```
  // We keep the dtypes to be relatively consistent to reduce the amount code generated by operators.
  // This can also help users find potential type conversion problems.
  // Specifically, when two dtypes are inconsistent, the int or lower bit one will be upcasted to match
  // another. E.g. fp16 + fp32 --> fp32 + fp32; fp32 + int16 --> fp32 + fp32.
```

File: src/tir/op/op.cc
Comment (comaniac) [inline src/tir/op/op.cc:130]: Looks like you will still add a cast to RHS even their bits are already the same?

```suggestion
    if (lhs.dtype().bits() < rhs.dtype().bits()) {
      lhs = cast(rhs.dtype(), lhs);
    } else if (lhs.dtype().bits() > rhs.dtype().bits()) {
      rhs = cast(lhs.dtype(), rhs);
    }
```

File: tests/python/unittest/test_tir_base.py
Comment (comaniac) [inline tests/python/unittest/test_tir_base.py:53]: Will this test fail (at all combinations of lhs_types and rhs_types except for `(float32, float32)`) without the changes in `op.cc`?

File: src/tir/op/op.cc
Comment (AndrewZhaoLuo) [reply src/tir/op/op.cc]: I've reorganized comments

File: src/tir/op/op.cc
Comment (AndrewZhaoLuo) [reply src/tir/op/op.cc]: If they have the same bits then it should hit the return statement above.

That being said your code is probably easier to read so I will replace it.

Done!

File: tests/python/unittest/test_tir_base.py
Comment (AndrewZhaoLuo) [reply tests/python/unittest/test_tir_base.py:53]: It should work except when mixing fp32 and fp16. I have tested the old version with fp32 and fp16 and it indeed fails.

The old code does do conversions between floating point types and integer types as expected.

**Comment by AndrewZhaoLuo** [comment]
Thanks for the review@comaniac PTAL

**Comment by AndrewZhaoLuo** [comment]
By the way, please don't merge this anytime soon. I want to get some input from some others like @jroesch or @junrushao1994

**Review by comaniac — APPROVED** [review]
LGTM. Leave it to @jroesch @junrushao1994

**Comment by jroesch** [comment]
As long as we don't implicitly downcast I think this should be fine, can you test behavior of storing both an i32/u32 computation in i16/u16, and the other way around?

**Comment by AndrewZhaoLuo** [comment]
> As long as we don't implicitly downcast I think this should be fine, can you test behavior of storing both an i32/u32 computation in i16/u16, and the other way around?

This will fail since this PR only upcasts one arg. in binary operations only. Do you want me to also upcast assignment for floating point types to be consistent?

**Comment by jroesch** [comment]
> > As long as we don't implicitly downcast I think this should be fine, can you test behavior of storing both an i32/u32 computation in i16/u16, and the other way around?
> 
> This will fail since this PR only upcasts at most one arg. in binary operations only (e.g. addition). Do you want me to also upcast assignment for floating point types to be consistent? @jroesch

I meant we should add negative tests so that if someone later comes along to modify the behavior we have clearly written down what should pass and what should fail.

**Comment by AndrewZhaoLuo** [comment]
> > > As long as we don't implicitly downcast I think this should be fine, can you test behavior of storing both an i32/u32 computation in i16/u16, and the other way around?
> > 
> > 
> > This will fail since this PR only upcasts at most one arg. in binary operations only (e.g. addition). Do you want me to also upcast assignment for floating point types to be consistent? @jroesch
> 
> I meant we should add negative tests so that if someone later comes along to modify the behavior we have clearly written down what should pass and what should fail.

Oh ok. Done.

**Comment by AndrewZhaoLuo** [comment]
PTAL @jroesch

**Comment by jroesch** [comment]
LGTM, just get it green and we are gtg

**Review by jroesch — APPROVED** [review]

**Comment by comaniac** [comment]
Thanks @AndrewZhaoLuo @jroesch

**Pull request merged**
